File: task-based-experiment/benchmarks/600.opentelemetry/610.dynamic-html-otel/python/function.py
from datetime import datetime                                                   
from random import sample  
from os import path
from time import time                                                           
import os
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.resources import Resource

from jinja2 import Template

SCRIPT_DIR = path.abspath(path.join(path.dirname(__file__)))


resource = Resource(attributes={
    "service.name": "610.dynamic-html-opentelemetry"
})

trace.set_tracer_provider(TracerProvider(resource=resource))
tracer = trace.get_tracer("function")

# Setup the OTLP exporter to send traces to the collector using HTTP
otlp_exporter = OTLPSpanExporter(
    endpoint="http://192.168.1.101:4317",
    insecure=True
)

# Setup the span processor with the exporter
span_processor = SimpleSpanProcessor(otlp_exporter)
trace.get_tracer_provider().add_span_processor(span_processor)


def handler(event):
    with tracer.start_as_current_span("handler") as span:
        # start timing
        name = event.get('username')
        size = event.get('random_len')
        cur_time = datetime.now()
        
        # Add attributes to the span
        span.set_attribute("username", name)
        span.set_attribute("random_len", size)
        
        random_numbers = sample(range(0, 1000000), size)
        template_path = os.path.join(SCRIPT_DIR, 'templates', 'template.html')
        
        # Read and render template
        with open(template_path, 'r') as file:
            template = Template(file.read())
        
        html = template.render(username=name, cur_time=cur_time, random_numbers=random_numbers)
        
        # Set additional attributes if needed
        span.set_attribute("template_path", template_path)
        span.set_attribute("render_time", str(cur_time))
        
        # end timing and return result
        return {'result': html}

File: task-based-experiment/benchmarks/600.opentelemetry/610.dynamic-html-otel/nodejs/function.js
const Mustache = require('mustache');
const fs = require('fs');
const path = require('path');
const { NodeTracerProvider } = require('@opentelemetry/node');
const { SimpleSpanProcessor } = require('@opentelemetry/tracing');
const { CollectorTraceExporter } = require('@opentelemetry/exporter-collector-grpc');
const { trace } = require('@opentelemetry/api');
const { Resource } = require('@opentelemetry/resources');

const provider = new NodeTracerProvider({
  resource: new Resource({
    'service.name': '610.dynamic-html-opentelemetry-nodejs',
  }),
});
const exporter = new CollectorTraceExporter({
  url: 'http://192.168.1.101:4317',
});
provider.addSpanProcessor(new SimpleSpanProcessor(exporter));
provider.register();

const tracer = trace.getTracer('tracer');

function random(b, e) {
  return Math.round(Math.random() * (e - b) + b);
}

exports.handler = async function (event) {
  return tracer.startActiveSpan('handler', { attributes: { 'handler.event.username': event.username, 'handler.event.random_len': event.random_len } }, async (span) => {

    var random_numbers = new Array(event.random_len);
    for (var i = 0; i < event.random_len; ++i) {
      random_numbers[i] = random(0, 100);
    }

    var input = {
      cur_time: new Date().toLocaleString(),
      username: event.username,
      random_numbers: random_numbers,
    };
    span.setAttribute('username', event.username)

    var file = path.resolve(__dirname, 'templates', 'template.html');
    return new Promise((resolve, reject) => {
      fs.readFile(file, 'utf-8', (err, data) => {
        if (err) {
          span.recordException(err);
          span.setStatus({ code: 2, message: 'File read error' });
          span.setAttribute('error', true);
          span.end();
          reject(err);
        } else {
          const rendered = Mustache.render(data, input);
          span.setAttribute('rendered.length', rendered.length);
          span.end();
          resolve(rendered);
        }
      });
    });
  });
};
